tests/phpunit/tests/blocks/block.php

```diff
@@ -394,4 +411,127 @@ function test_passes_content_to_render_callback() {
+
+		add_filter( 'pre_render_block', array( $this, 'filter_pre_render_block' ), 10, 2 );
+
+		$parsed_blocks = parse_blocks( '<!-- wp:outer --><!-- wp:skip /--> How are you?<!-- /wp:outer -->' );
+		$parsed_block  = $parsed_blocks[0];
+		$context       = array();
+		$block         = new WP_Block( $parsed_block, $context, $this->registry );
+
+		$rendered_content = $block->render();
+
+		remove_filter( 'pre_render_block', array( $this, 'filter_pre_render_block' ) );
+
+		$this->assertSame( 'Hello world! How are you?', $rendered_content );
+	}
+
+	/**
+	 * @ticket 51612
+	 */
+	function test_applies_render_block_data_filter() {
+		$this->registry->register(
+			'core/wrapper',
+			array(
+				'attributes'      => array(
+					'tag' => array(
+						'type' => 'string',
+					),
+				),
+				'render_callback' => function( $block_attributes, $content ) {
+					return sprintf(
+						'<%1$s>%2$s</%1$s>',
+						$block_attributes['tag'],
+						$content
+					);
+				},
+			)
+		);
+
+		add_filter( 'render_block_data', array( $this, 'filter_render_block_data' ), 10, 2 );
+
+		$parsed_blocks = parse_blocks( '<!-- wp:wrapper {"tag":"outer"} --><!-- wp:wrapper {"tag":"inner"} -->Hello!<!-- /wp:wrapper --><!-- /wp:wrapper -->' );
+		$parsed_block  = $parsed_blocks[0];
+		$context       = array();
+		$block         = new WP_Block( $parsed_block, $context, $this->registry );
+
+		$rendered_content = $block->render( array( 'filter_data' => true ) );
+
+		remove_filter( 'render_block_data', array( $this, 'filter_render_block_data' ) );
+
+		$this->assertSame( '<outer-filtered><inner-filtered>Hello!</inner-filtered></outer-filtered>', $rendered_content );
+	}
+
+	/**
+	 * @ticket 51612
+	 */
+	function test_applies_render_block_context_filter() {
+		$this->registry->register(
+			'core/provider',
+			array(
+				'attributes'       => array(
+					'recordId' => array(
+						'type' => 'number',
+					),
+				),
+				'uses_context'     => array( 'core/recordId' ),
+				'provides_context' => array(
+					'core/recordId' => 'recordId',
+				),
+			)
+		);
+		$this->registry->register(
+			'core/consumer',
+			array(
+				'uses_context'    => array( 'core/recordId' ),
+				'render_callback' => function( $block_attributes, $content, $block ) {
+					return sprintf( 'Record ID: %d ', $block->context['core/recordId'] );
+				},
+			)
+		);
+
+		add_filter( 'render_block_context', array( $this, 'filter_render_block_context' ), 10, 2 );
+
+		$parsed_blocks = parse_blocks( '<!-- wp:consumer /--><!-- wp:provider {"recordId":20} --><!-- wp:consumer /--><!-- /wp:provider -->' );
+		$context       = array( 'core/recordId' => 10 );
+
+		$rendered_content = '';
+
+		foreach ( $parsed_blocks as $parsed_block ) {
+			$block = new WP_Block( $parsed_block, $context, $this->registry );
+
+			$rendered_content .= $block->render( array( 'filter_data' => true ) );
+		}
+
+		remove_filter( 'render_block_context', array( $this, 'filter_render_block_context' ) );
```

Comment:
Nit: No need to remove filters/actions in the test, they're reset after each test as part of the default `tearDown()`.

Reply:
Oh cool! Is it better to be explicit though?